packages/node/src/indexer/fetch.service.ts

```diff
@@ -100,41 +214,99 @@ export class FetchService implements OnApplicationShutdown {
   latestProcessed(height: number): void {
     this.latestProcessedHeight = height;
   }
-
   async startLoop(initBlockHeight: number): Promise<void> {
+    await Promise.all([
+      this.setNextBlockBuffer(initBlockHeight),
+      this.setBlockBuffer(),
+    ]);
+  }
+
+  async setNextBlockBuffer(initBlockHeight: number): Promise<void> {
     if (isUndefined(this.latestProcessedHeight)) {
       this.latestProcessedHeight = initBlockHeight - 1;
     }
     await this.fetchMeta(initBlockHeight);
     // eslint-disable-next-line no-constant-condition
+    let startBlockHeight;
+    startBlockHeight = initBlockHeight;
+
+    while (!this.isShutdown) {
+      if (this.useDictionary) {
+        try {
+          const dictionary = await this.getDictionary(startBlockHeight);
+          const batchBlocks = this.dictionaryService.getBlockBatch(dictionary);
+          //TODO
+          // const specVersionMap = this.dictionaryService.getSpecVersionMap(dictionary);
+          const metaData = dictionary._metadata;
+          if (
+            metaData.chain !== this.api.runtimeChain.toString() ||
+            metaData.specName !== this.api.runtimeVersion.specName.toString() ||
+            metaData.genesisHash !== this.api.genesisHash.toString()
+          ) {
+            logger.warn(
+              `Dictionary is not matching with this network endpoint`,
+            );
+            this.useDictionary = false;
+            continue;
+          }
+          if (metaData.lastProcessedHeight < startBlockHeight) {
+            logger.warn(
+              `Dictionary indexed block is behind current indexing block height`,
+            );
+            this.useDictionary = false;
+            continue;
+          }
+          if (isUndefined(batchBlocks)) {
+            await delay(1);
+            continue;
+          }
+          if (batchBlocks.length !== 0) {
+            this.nextBlockBuffer.putAll(batchBlocks);
+            this.latestPreparedHeight = batchBlocks[batchBlocks.length - 1];
```

Comment:
`latestPreparedHeight` doesn't change until the block is fetched into blockBuffer